Introduce Nexus.Logger abstraction for logging

ZacSweers opened the pull request:
This reworks some of the logging in Nexus to allow for using Gradle's `ProgressLogger`, which offers a richer (and more cooperative) logging mechanism per task. I'm not sure how these closing tasks are tested here, but in another project it looks like this: https://github.com/slackhq/slack-gradle-plugin/pull/247

This also tweaks the update loop to update the progress char every 200ms and opportunistically lowers the repository update interval to 5 seconds, as 10 seems quite long to wait. Can revert those as needed (or make them configurable?)

One thing I can't quite figure out is the kludged text at the end, but that may just be a fact of life with carriage returns


https://user-images.githubusercontent.com/1361086/219898618-d6abfb56-0931-49cb-ac95-13ed6b8b4801.mov

ZacSweers commented:
Going to try to get an example using mavenLocal to demo

gabrielittner commented on nexus/src/main/kotlin/com/vanniktech/maven/publish/nexus/Nexus.kt:
```diff
@@ -274,6 +280,36 @@ class Nexus(
+      }
+      override fun completed(status: String, failed: Boolean) {
+        if (failed) {
+          System.err.println("Completed with errors: $status")
+        } else {
+          println("Completed: $status")
```
Can you try adding a \n at the beginning of both of these? That should fix the missing new line for the completed releasing message at the end, it's probably coming from the progress updates never ending in a new line

ZacSweers replied:
[b876a23](https://github.com/vanniktech/gradle-maven-publish-plugin/pull/525/commits/b876a23bf04fb4412aba4191189a1583868dbdbc)

gabrielittner reviewed (commented):
Some small comments, otherwise looks good and is a really cool change. The timeo

For testing: unfortunately the sonatype stuff has no automated tests so far, I didn't get to that yet.

> This also tweaks the update loop to update the progress char every 200ms and opportunistically lowers the repository update interval to 5 seconds, as 10 seems quite long to wait. Can revert those as needed (or make them configurable?)

Lowering the update to 5s should be fine. 

Do you know if Gradle exposes an API to check if the console is interactive? An update 200ms might be too much for CI and then we could autonatically adjust that.

ZacSweers commented:
Sorry for the delay on this, got swamped with kotlinconf and some other work stuff. Addressing comments now.

> Do you know if Gradle exposes an API to check if the console is interactive?

It looks like there is via `project.gradle.startParams.consoleOutput`. Will look into it

ZacSweers commented on nexus/src/main/kotlin/com/vanniktech/maven/publish/nexus/Nexus.kt:
```diff
@@ -160,11 +160,11 @@ class Nexus(
 
-    waitForClose(repositoryId)
+    waitForClose(repositoryId, logger)
     println("Repository $repositoryId closed")
```
Part of [55bcb21](https://github.com/vanniktech/gradle-maven-publish-plugin/pull/525/commits/55bcb2192a2b9ec37981b89c6cf6e458aebb66be)

ZacSweers commented on nexus/src/main/kotlin/com/vanniktech/maven/publish/nexus/Nexus.kt:
```diff
@@ -138,12 +138,12 @@ class Nexus(
-        waitForClose(stagingRepository.repositoryId)
+        waitForClose(stagingRepository.repositoryId, logger)
       } else {
         println("Repository $repositoryId already closed")
```
part of [55bcb21](https://github.com/vanniktech/gradle-maven-publish-plugin/pull/525/commits/55bcb2192a2b9ec37981b89c6cf6e458aebb66be)

ZacSweers commented on nexus/src/main/kotlin/com/vanniktech/maven/publish/nexus/Nexus.kt:
```diff
@@ -138,12 +138,12 @@ class Nexus(
   }
 
-  private fun closeStagingRepository(stagingRepository: Repository) {
+  private fun closeStagingRepository(stagingRepository: Repository, logger: Logger) {
```
started in [9b1ce5d](https://github.com/vanniktech/gradle-maven-publish-plugin/pull/525/commits/9b1ce5d3ee50d9cf41dd160427c8bbca4c731b54)

ZacSweers commented:
I have to abandon this PR actually, as I don't believe any of this is compatible with workers as `ProgressLoggerFactory` is not injectable to workers. They're fairly limited unfortunately. That changed on main since I started this, apologies for the noise
